libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsDownloadTrigger.kt

```diff
@@ -0,0 +1,149 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesExtra
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.directions.session.RoutesUpdatedResult
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.utils.internal.Time
+import com.mapbox.navigation.utils.internal.ifNonNull
+
+/**
+ * Class that is responsible for listening to relevant updates and triggering
+ * voice instructions predownloading when needed.
+ * Register and unregister with [MapboxNavigation.registerVoiceInstructionsTriggerObserver] and
+ * [MapboxNavigation.unregisterVoiceInstructionsTriggerObserver].
+ */
+class VoiceInstructionsDownloadTrigger internal constructor(
```

Comment:
```suggestion
class VoiceInstructionsDownloader internal constructor(
```

Am nitpicking the name, because I've regretted the name of ReplayProgressObserver. Observers are interfaces not services.

Instead of implementing both the RouteProgressObserver and the RoutesObserver. I think it should be a `MapboxNavigationObserver`. That would allow us to move the function `mapboxNavigation.registerVoiceInstructionsTriggerObserver` into an `onAttached`. Makes it easier to integrate with MapboxNavigationApp. It also gives the voice instruction download the capability to register new observers (maybe in the future, we want to use location updates to handle voice instruction downloads).

Reply:
> Am nitpicking the name, because I've regretted the name of ReplayProgressObserver. Observers are interfaces not services.

I'm not sure if removing the "trigger" word would make it more clear. It actually triggers the download, not performs the downloading itself.

> Instead of implementing both the RouteProgressObserver and the RoutesObserver. I think it should be a MapboxNavigationObserver. That would allow us to move the function mapboxNavigation.registerVoiceInstructionsTriggerObserver into an onAttached. Makes it easier to integrate with MapboxNavigationApp. It also gives the voice instruction download the capability to register new observers (maybe in the future, we want to use location updates to handle voice instruction downloads).

But how will it work for those users who don't use `MapboxNavigationApp` then?

Reply:
> But how will it work for those users who don't use MapboxNavigationApp then?

```kotlin
val downloader = VoiceInstructionsDownloader()
downloader.onAttached(mapboxNavigation)
// later in onDestory
downloader.onDetached(mapboxNavigation)
```

Reply:
That makes sense.
Thanks, I'll change that.

Reply:
> I'm not sure if removing the "trigger" word would make it more clear. It actually triggers the download, not performs the downloading itself.

Sorry my comment was confusing to bring up the "Observer", clearly unrelated. I was looking for an explanation for why "Trigger" feels wrong to me. Also "Downloader" feels wrong to me 🤔 . So am only sharing naming ideas to see if something feels better. 

"Prefetch" is a common computing term, what do you think about "VoiceInstructionPrefetcher"? I like it when services have a single word to describe what they do.

Reply:
Leaving the name up to you am only sharing opinions. Resolving this thread